incubator/hnc/internal/reconcilers/object.go

```diff
@@ -332,11 +333,13 @@ func (r *ObjectReconciler) syncObject(ctx context.Context, log logr.Logger, inst
 
 	// Find the source object of the same name in the ancestors from top down to
 	// see if there's a conflicting source.
-	srcInst := r.Forest.Get(inst.GetNamespace()).GetSource(r.GVK, inst.GetName())
+	srcInst := r.Forest.Get(inst.GetNamespace()).GetAncestorSourceObject(r.GVK, inst.GetName())
 
 	// The object is a source without conflict if a copy of the source is not
-	// found in the forest or itself is found.
-	if srcInst == nil || srcInst.GetNamespace() == inst.GetNamespace() {
+	// found in the forest or itself is found. There won't be any conflict for
+	// types with sync modes other than 'Propagate' either.
+	mode := r.Forest.GetTypeSyncer(inst.GroupVersionKind()).GetMode()
+	if srcInst == nil || srcInst.GetNamespace() == inst.GetNamespace() || mode != api.Propagate {
```

Comment:
How could `srcInst.GetNamespace() == inst.GetNamespace()` today? Can't we get rid of this clause?

I also wonder if this code would now be written more naturally in the reverse (I had a bit of a hard time following the logic):

```
if srcInst != nil && mode == api.Propagate {
   // Since this is a conflict, overwrite
   return r.syncPropagated(ctx, log, inst)
}

// This object is not being overwritten; register the new source object and don't take any other action.
r.syncSource(ctx, log, inst)
return actionNop, nil
```

Or maybe even combine both calls in syncPropagated to one in this function by changing `hasPropagatedLabel` to `isPropagated`? That way you don't then have to "rediscover" the source object inside `syncPropagated` as we do today. E.g.:

```
if srcInst, isPropagated := r.isPropagated(inst); isPropagated {
  // srcInst can be nil if the source was removed
  return r.syncPropagated(ctx, log, srcInst, inst)
}
r.syncSource(ctx, log, inst)
return actionNop, nil
```
```

Reply:
sgtm, reversed the logic!